File: proto/control-api-mock/src/room.rs
use std::collections::HashMap;

use crate::member::Member;

use serde::{Deserialize, Serialize};

/// [Control API]'s `Room` representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
#[derive(Debug, Deserialize, Serialize)]
pub struct Room {
    /// ID of this [`Room`].
    #[serde(skip_deserializing)]
    pub id: String,

    /// Pipeline of this [`Room`].
    pub pipeline: HashMap<String, RoomElement>,
}

/// Element of [`Room`]'s pipeline.
#[derive(Debug, Deserialize, Serialize)]
#[serde(tag = "kind")]
pub enum RoomElement {
    Member(Member),
}

File: flutter/example/lib/control_api/entities/room.dart
// ignore_for_file: depend_on_referenced_packages

import 'package:json_annotation/json_annotation.dart';

import 'member.dart';

part 'room.g.dart';

/// [Control API]'s [Room] representation.
///
/// [Control API]: https://tinyurl.com/yxsqplq7
@JsonSerializable()
class Room {
  /// ID of this [Room].
  String id;

  /// Pipeline of this [Room].
  Map<String, Member> pipeline;

  Room(this.id, this.pipeline);

  factory Room.fromJson(Map<String, dynamic> json) {
    json.remove('kind');
    return _$RoomFromJson(json);
  }

  Map<String, dynamic> toJson() {
    var res = _$RoomToJson(this);
    res.addAll({'kind': 'Room'});
    return res;
  }
}
